src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -257,19 +433,84 @@ private static void BuildSymbolInformationFromOperations(OperationAnalysisContex
             }
         }
 
-        private static bool OperationUsesPreviewFeatures(OperationAnalysisContext context, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, INamedTypeSymbol previewFeatureAttributeSymbol, [NotNullWhen(true)] out ISymbol? symbol)
+        private static bool SymbolIsAnnotatedOrUsesPreviewTypes(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, INamedTypeSymbol previewFeatureAttributeSymbol, [NotNullWhen(true)] out ISymbol? referencedPreviewSymbol)
+        {
+            if (SymbolIsAnnotatedAsPreview(symbol, requiresPreviewFeaturesSymbols, previewFeatureAttributeSymbol))
+            {
+                referencedPreviewSymbol = symbol;
+                return true;
+            }
+
+            if (SymbolContainsGenericTypesWithPreviewAttributes(symbol, requiresPreviewFeaturesSymbols, previewFeatureAttributeSymbol, out referencedPreviewSymbol))
+            {
+                return true;
+            }
+
+            referencedPreviewSymbol = null;
+            return false;
+        }
+
+        private static bool OperationUsesPreviewFeatures(OperationAnalysisContext context, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, INamedTypeSymbol previewFeatureAttributeSymbol, [NotNullWhen(true)] out ISymbol? referencedPreviewSymbol)
         {
             IOperation operation = context.Operation;
             ISymbol containingSymbol = context.ContainingSymbol;
-            if (OperationUsesPreviewFeatures(containingSymbol, requiresPreviewFeaturesSymbols, previewFeatureAttributeSymbol) ||
-                OperationUsesPreviewFeatures(containingSymbol.ContainingSymbol, requiresPreviewFeaturesSymbols, previewFeatureAttributeSymbol))
+            if (SymbolIsAnnotatedAsPreview(containingSymbol, requiresPreviewFeaturesSymbols, previewFeatureAttributeSymbol))
             {
-                symbol = null;
+                referencedPreviewSymbol = null;
                 return false;
```

Comment:
> Is this the logic for guarding the call site?

Yup.

> Doesn't look like work for all cases, but I might not understand completely

Why not? Can you think of an example where this might not work? `containinSymbol` here is `context.Operation.ContainingSymbol`, so it is the parent to the current operation. So I'm checking to see if the parent to the current location is marked as preview, and returning false if it is. I don't see what I'm missing :/

Reply:
> Can you think of an example where this might not work?

No, thinking again about the scenario i was thinking has no issue. This guard looks good for Operation analysis